core-rust/state-manager/src/protocol/protocol_state.rs

```diff
@@ -44,37 +44,63 @@ impl ProtocolUpdateExecutor {
         }
     }
 
+    /// Executes any remaining parts of the currently-effective protocol update.
+    /// This method is meant to be called during the boot-up, to support resuming after a restart.
+    pub fn execute_remaining_protocol_update_actions(&self) {
+        match ProtocolUpdateProgress::resolve(self.database.lock().deref()) {
+            ProtocolUpdateProgress::UpdateInitiatedButNothingCommitted {
+                protocol_version_name,
+            } => {
+                info!(
+                    "Starting a {} protocol update execution",
+                    protocol_version_name
+                );
+                self.execute_protocol_update(&protocol_version_name, 0);
+            }
+            ProtocolUpdateProgress::UpdateInProgress {
+                protocol_version_name,
+                last_batch_idx,
+            } => {
+                let next_batch_idx = last_batch_idx.checked_add(1).unwrap();
+                info!(
+                    "Resuming a {} protocol update execution from batch idx {}",
+                    protocol_version_name, next_batch_idx
+                );
+                self.execute_protocol_update(&protocol_version_name, next_batch_idx);
+            }
+            ProtocolUpdateProgress::NotUpdating => {} // No protocol update in progress
+        }
+    }
+
+    /// Executes all transactions for the given new protocol update.
+    /// This method is meant to be called by the consensus process.
+    pub fn execute_new_protocol_update(&self, new_protocol_version: &ProtocolVersionName) {
+        info!("Executing {} protocol update", new_protocol_version);
+        self.execute_protocol_update(new_protocol_version, 0)
+    }
+
     /// Executes the (remaining part of the) given protocol update's transactions.
-    pub fn execute_protocol_update(&self, new_protocol_version: &ProtocolVersionName) {
-        let overrides = self
-            .protocol_update_content_overrides
-            .get(new_protocol_version);
-        let action_provider = resolve_update_definition_for_version(new_protocol_version)
-            .unwrap_or_else(|| panic!("{}", new_protocol_version.as_str().to_string()))
+    fn execute_protocol_update(&self, protocol_version: &ProtocolVersionName, from_batch_idx: u32) {
```

Comment:
I understand that we want more consistent naming...
However, `batch_idx` has kinda won already, since this name has been used in the Core API for quite some time now (see `ProtocolUpdateLedgerProofOrigin`).

Out of necessity, I accepted that "protocol update actions" are indexed by "batch idx" :/

So, if we want consistency here, then I'd probably have to rename the concept of "PU action" (which is "either a batch of flash transactions, or a single complete test scenario") back to "PU batch".
It is not a big deal, and not very wrong conceptually (because a test scenario is actually also "some batch of transactions" really), but I'd have to creatively resolve name-clashes with some Engine's structs (like `ProtocolUpdateBatch`).

What do you think?

Reply:
No problem. I've seen both action and batch being used in the PR. Not sure if we need both concepts.

Reply:
That's true, it is a single underlying concept.
So I rename it back to "batch generator" everywhere in the newest commit. I had to apply some disambiguation in naming (i.e. "Node batch" vs "Engine batch") and explained it in a rustdoc 👍